### PR TITLE
UCP/API: request based tag receive

### PR DESCRIPTION
**Pull request by brminich**
- Added ucp_tag_recv_nbr, ucp_request_test and ucp_context_query calls
- Tag matching tests updated to run with new API as well

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2277/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1605/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2278/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1606/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -338,6 +338,23 @@ typedef struct ucp_params {
...
+ * the particular context.
+ */
+typedef struct ucp_context_attr {
+    /**
+     * Size of UCP non-blocking request. When pre-allocated request is used
+     * (e.g. in @ref ucp_tag_recv_nbr) it should have enough space to fit
+     * UCP request data, which is defined by this value.
+     */
+    size_t                request_size;
+}ucp_context_attr_t;
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: missing space

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -338,6 +338,23 @@ typedef struct ucp_params {
 
 /**
  * @ingroup UCP_CONTEXT
+ * @brief Context attributes.
+ *
+ * The structure defines the attributes which characterize
+ * the particular context.
+ */
+typedef struct ucp_context_attr {
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: let's also add "field_mask" parameter for compatibility - so user could specify which fields in its context attributes should be returned by this function

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -1157,6 +1193,39 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
...
+ *
+ * @param [in]  worker      UCP worker that is used for the receive operation.
+ * @param [in]  buffer      Pointer to the buffer to receive the data to.
+ * @param [in]  count       Number of elements to receive
+ * @param [in]  datatype    Datatype descriptor for the elements in the buffer.
+ * @param [in]  tag         Message tag to expect.
+ * @param [in]  tag_mask    Bit mask that indicates the bits that are used for
+ *                          the matching of the incoming tag
+ *                          against the expected tag.
+ * @param [in]  req         Request handle allocated by the user. It should have
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: let's make it more clear - "there should be at least request_size bytes of available space before *req"

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -1607,6 +1676,26 @@ int ucp_request_is_completed(void *request);
...
  * @ingroup UCP_COMM
+ * @brief Check the status of non-blocking request.
+ *
+ * This routine checks the state of the request and returns its current status.
+ * Any value different from UCS_INPROGRESS means that request is in a completed
+ * state.
+ *
+ * @param [in]  request     Non-blocking request to check.
+ *
+ * @param [out] info        If request is in completed state, it is
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: currently "info" is set to "garbage" in case of send request. Is this what we want? if so need to document this/

File: src/ucp/core/ucp_request.c [hunk 5]
```diff
@@ -20,6 +20,20 @@ int ucp_request_is_completed(void *request)
     return !!(req->flags & UCP_REQUEST_FLAG_COMPLETED);
 }
 
+ucs_status_t ucp_request_test(void *request, ucp_tag_recv_info_t *info)
+{
+  ucp_request_t *req = (ucp_request_t*)request - 1;
+
+  if (req->flags & UCP_REQUEST_FLAG_COMPLETED)
+  {
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c]: code style + indentation

File: src/ucp/core/ucp_request.c [hunk 6]
```diff
@@ -20,6 +20,20 @@ int ucp_request_is_completed(void *request)
...
+    info->length = req->recv.info.length;
+    info->sender_tag = req->recv.info.sender_tag;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c]: note these are valid only for receive requests

Comment (yosefe) [inline src/ucp/core/ucp_request.c]: can also do *info = req->recv.info

File: src/ucp/core/ucp_request.c [hunk 7]
```diff
@@ -52,6 +66,8 @@ static void ucp_worker_request_init_proxy(ucs_mpool_t *mp, void *obj, void *chun
     ucp_context_h context = worker->context;
     ucp_request_t *req = obj;
 
+    req->status = UCS_INPROGRESS;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c]: why is this needed?

File: src/ucp/core/ucp_request.inl [hunk 8]
```diff
@@ -54,12 +55,18 @@ ucp_request_complete_recv(ucp_request_t *req, ucs_status_t status,
     ucs_trace_data("completing recv request %p (%p) stag 0x%"PRIx64" len %zu, %s",
                    req, req + 1, info->sender_tag, info->length,
                    ucs_status_string(status));
-    req->recv.cb(req + 1, status, info);
+
+    if (req->flags & UCP_REQUEST_FLAG_EXTERNAL) {
+        req->flags |= UCP_REQUEST_FLAG_COMPLETED;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl]: this seems weird - shouldn't the "COMPLETED" flag be already set by the caller?
and shouldn't we call the callback for external requests as well (in case it's != NULL)?
IMHO the only difference for EXTERNAL request should be that it's not passed to ucp_request_put, all rest of handling should be the same.

File: src/ucp/core/ucp_request.inl [hunk 9]
```diff
@@ -54,12 +55,18 @@ ucp_request_complete_recv(ucp_request_t *req, ucs_status_t status,
...
+        req->status = status;
+    } else {
+        req->recv.cb(req + 1, status, info);
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl]: seems like doing ucp_request_test on an internal request would not return the correct status?

File: src/ucp/tag/tag_recv.c [hunk 10]
```diff
@@ -107,42 +114,30 @@ ucp_tag_recv_request_completed(ucp_request_t *req, ucp_tag_recv_callback_t cb,
...
+                  buffer, count, tag, tag_mask);
 
     /* First, search in unexpected list */
     status = ucp_tag_search_unexp(worker, buffer, count, datatype, tag,
                                   tag_mask, req, &req->recv.info, &save_rreq);
-    if (status != UCS_INPROGRESS) {
-        ucp_tag_recv_request_completed(req, cb, status, &req->recv.info,
-                                       "recv_nb");
-    } else if (save_rreq) {
+    if (save_rreq && status == UCS_INPROGRESS) {
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: this change add up to 2 branches to fast path - in expected case, instead of checking once and completing the request, there are 2 more checks here. Need to check for expected case first.

File: src/ucp/tag/tag_recv.c [hunk 11]
```diff
@@ -151,9 +146,65 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
...
+    UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX, "ucp_tag_recv_nbr",
+                          req, ucp_contig_dt_length(datatype, count));
+
+    req->flags   |= UCP_REQUEST_FLAG_EXTERNAL;
+    req->recv.cb  = NULL;
+    status = ucp_tag_recv_common(worker, buffer, count, datatype, tag,
+                                 tag_mask, req);
+
+    if (status != UCS_INPROGRESS) {
+        UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX,
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: need to also add debug trace here and call user callback if it's != NULL - maybe call ucp_tag_recv_completed() with some compile-time indication that the request is external?

File: src/ucp/tag/tag_recv.c [hunk 12]
```diff
@@ -151,9 +146,65 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
...
+    ucp_request_t  *req = (ucp_request_t *)request - 1;
+    ucs_status_t status;
+
+    ucp_tag_recv_request_init(req, worker, buffer, count, datatype);
+
+    UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX, "ucp_tag_recv_nbr",
+                          req, ucp_contig_dt_length(datatype, count));
+
+    req->flags   |= UCP_REQUEST_FLAG_EXTERNAL;
+    req->recv.cb  = NULL;
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: don't we want to allow user to set callback?

File: src/ucp/core/ucp_request.inl (hunk 9 above)
Comment (brminich) [inline src/ucp/core/ucp_request.inl]: No, status is set in ucp_request_put. It is covered by new tests

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (brminich) [inline src/ucp/api/ucp.h]: There is a note in this comment that info is for recv reqs only. I'll mention garbage explicitly

File: src/ucp/core/ucp_request.inl (hunk 8 above)
Comment (brminich) [inline src/ucp/core/ucp_request.inl]: COMPLETED flag is set in ucp_request_put, that's why we set it separately.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2279/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2280/ for details.

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (yosefe) [inline src/ucp/api/ucp.h]: yes - user would expect that for recv request this would not be changed, or event info=NULL could be passed

File: src/ucp/core/ucp_request.inl (hunk 8 above)
Comment (yosefe) [inline src/ucp/core/ucp_request.inl]: in that case i think need same flow for external and internal requests, and just avoid setting UCP_REQUEST_FLAG_RELEASED for external requests, so ucs_mpool_put would not be called

File: test/gtest/ucp/test_ucp_tag.cc [hunk 13]
```diff
@@ -154,6 +220,24 @@ ucs_status_t test_ucp_tag::recv_b(void *buffer, size_t count, ucp_datatype_t dat
...
 }
 
+ucs_status_t test_ucp_tag::recv_req_b(void *buffer, size_t count, ucp_datatype_t datatype,
+                                      ucp_tag_t tag, ucp_tag_t tag_mask,
+                                      ucp_tag_recv_info_t *info)
+{
+    request *req = request_alloc();
+    ucs_status_t status = ucp_tag_recv_nbr(receiver().worker(), buffer, count,
+                                           datatype, tag, tag_mask, req);
+    if (status != UCS_OK && status != UCS_INPROGRESS) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag.cc]: code style

**Comment by brminich** [comment]
@shssf, could you please look thru as well? (used your suggestions for tests update)

File: src/ucp/core/ucp_request.h [hunk 14]
```diff
@@ -60,6 +61,7 @@ typedef struct ucp_send_state {
  * Request in progress.
  */
 struct ucp_request {
+    ucs_status_t                  status;  /* Operation status */
```
Comment (shssf) [inline src/ucp/core/ucp_request.h]: Do we need this field as the first one in the structure? It might displace usefully data from the first cache line.

Comment (brminich) [inline src/ucp/core/ucp_request.h]: I've made several experiments/micro-benchmarking with it, and found no performance implications from moving this into the end of struct. So finally I added it here close to the flag.

File: src/ucp/api/ucp.h [hunk 15]
```diff
@@ -338,6 +350,31 @@ typedef struct ucp_params {
...
+     */
+    size_t                request_size;
+
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref ucp_context_attr_field.
+     * Fields not specified in this mask will be ignored.
+     * Provides ABI compatibility with respect to adding new fields.
+     */
+    uint64_t                           field_mask;
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: pls indent fields on same column

File: src/ucp/core/ucp_request.h (hunk 14 above)
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: by how much does this field increase request size?

File: src/ucp/tag/tag_recv.c [hunk 16]
```diff
@@ -151,9 +147,63 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
...
+{
+    ucp_request_t  *req = (ucp_request_t *)request - 1;
+    ucs_status_t status;
+
+    ucp_tag_recv_request_init(req, worker, buffer, count, datatype);
+
+    UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX, "ucp_tag_recv_nbr",
+                          req, ucp_contig_dt_length(datatype, count));
+
+    req->flags   |= UCP_REQUEST_FLAG_EXTERNAL;
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: maybe could pass "extra_flags" to ucp_tag_recv_request_init() to avoid this additional |=

File: test/gtest/ucp/test_ucp_tag_match.cc [hunk 17]
```diff
@@ -15,6 +15,32 @@ using namespace ucs; /* For vector<char> serialization */
...
+        std::vector<ucp_test_param> tmp, tmp_ret;
+        ucp_test_param test_param;
+
+        tmp = ucp_test::enum_test_params(ctx_params, name, test_case_name, tls);
+        for (std::vector<ucp_test_param>::iterator it = tmp.begin(); it != tmp.end(); it++)
+        {
+            it->variant = false;
+            tmp_ret.push_back(*it);
+            it->variant = true;
+            tmp_ret.push_back(*it);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_match.cc]: does this mean we instantiate all UCP tests twice, even those which do not use the variant (e.g the tests without tag matching)?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2285/ for details.

File: test/gtest/ucp/test_ucp_tag_match.cc (hunk 17 above)
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_match.cc]: No, it does not. It is only applicable to test_ucp_tag_match, as we define this function there

File: src/ucp/tag/tag_recv.c (hunk 16 above)
Comment (brminich) [inline src/ucp/tag/tag_recv.c]: Can do that. Why would we need that?

Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: to avoid the additional |=operation

Comment (brminich) [inline src/ucp/tag/tag_recv.c]: I see, it can be done without extra branches, right

File: src/ucp/core/ucp_request.h (hunk 14 above)
Comment (brminich) [inline src/ucp/core/ucp_request.h]: it's 176 bytes now as it was before (due to alignment/padding). Have not investigated the data struct layout with tools lake pahole due to that.
If we add status to the end, it becomes somewhat 182 or 184

Comment (yosefe) [inline src/ucp/core/ucp_request.h]: good. so let's keep it in the beginning.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1610/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2286/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1611/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2287/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2288/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2289/ for details.

**Comment by yosefe** [comment]
memory leak:
```
223 ucs_status_t test_ucp_tag::recv_req_b(void *buffer, size_t count, ucp_datatype_t datatype,
224                                       ucp_tag_t tag, ucp_tag_t tag_mask,
225                                       ucp_tag_recv_info_t *info)
226 {
227     request *req = request_alloc();
```

**Comment by yosefe** [comment]
:+1:  just need to fix valgrind issue

**Comment by shssf** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2290/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1612/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2295/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1616/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1624/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2303/ for details.

**Comment by yosefe** [comment]
@shamisp can you pls review as well?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1625/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@MattBBaker, could you please post ORNL error?

**Comment by MattBBaker** [comment]
The /dev/shm error, happened because this test collided with another one, exhausted shm memory.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1627/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1630/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2310/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1635/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 18]
```diff
@@ -546,6 +583,25 @@ void ucp_cleanup(ucp_context_h context_p);
 
 
 /**
+ * @ingroup UCP_CONTEXT
+ * @brief Get attributes specific to a particular context.
+ *
+ * This routine fetches an information about the context.
+ *
+ * @warning An application cannot call any UCP routine
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Or before it is initialized ...

File: src/ucp/api/ucp.h [hunk 19]
```diff
@@ -1157,6 +1213,39 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
...
+ *                          be at least UCP request size bytes of available
+ *                          space before the @a req. The size of UCP request
+ *                          can be obtained by @ref ucp_context_query function.
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ */
+ucs_status_t ucp_tag_recv_nbr(ucp_worker_h worker, void *buffer, size_t count,
+                              ucp_datatype_t datatype, ucp_tag_t tag,
+                              ucp_tag_t tag_mask, void *req);
+
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Can we complete the request in-place ? is it useful option to have ?

File: src/ucp/api/ucp.h [hunk 20]
```diff
@@ -1607,6 +1696,26 @@ int ucp_request_is_completed(void *request);
...
+ * This routine checks the state of the request and returns its current status.
+ * Any value different from UCS_INPROGRESS means that request is in a completed
+ * state.
+ *
+ * @param [in]  request     Non-blocking request to check.
+ *
+ * @param [out] info        If request is in completed state, it is
+ *                          filled with the details about the message.
+ *
+ * @note The @p info parameter is relevant for receive operations only.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: You have specify what does it mean for send ? Do we leave it uninitialized or set it to NULL ?

**Comment by brminich** [comment]
@shamisp, please take a look at the latest update. 
BTW, I've removed the misleading warning, as it makes sense for functions which create/delete context only (it was just copied from ucp_cleanup).

**Comment by shamisp** [comment]
Looks good. 👍  . Please also check with MPICH folks.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2319/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1644/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@raffenet can we merge this?

File: src/ucp/api/ucp.h [hunk 21]
```diff
@@ -338,6 +350,31 @@ typedef struct ucp_params {
...
+     */
+    size_t                request_size;
+
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref ucp_context_attr_field.
+     * Fields not specified in this mask will be ignored.
+     * Provides ABI compatibility with respect to adding new fields.
+     */
+    uint64_t              field_mask;
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: missed this earlier - need to move field_mask to the beginning of the struct

**Comment by brminich** [comment]
@yosefe, Moved field_mask to the beginning + squashed commits

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1659/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1660/ for details (Mellanox internal link).

**Comment by MrBr-github** [comment]
bot:mellanox:retest

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2328/ for details.

**Comment by brminich** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2331/ for details.

**Comment by raffenet** [comment]
This looks good for usage in MPICH. :+1:
